contracts/vortex/CarbonVortex.sol

```diff
@@ -675,6 +679,10 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
         if (sourceAmount == 0) {
             revert InvalidTrade();
         }
+        // revert if unnecessary native token is received
+        if (_finalTargetToken != NATIVE_TOKEN && msg.value > 0) {
+            revert UnnecessaryNativeTokenReceived();
+        }
```

Comment:
You can put this in the `else` a few lines below, i.e.:
```
        if (_finalTargetToken == NATIVE_TOKEN) {
            if (msg.value < sourceAmount) {
                revert InsufficientNativeTokenSent();
            }
            payable(_transferAddress).sendValue(sourceAmount);
        } else {
            if (_finalTargetToken != NATIVE_TOKEN && msg.value > 0) {
                revert UnnecessaryNativeTokenReceived();
            }
            _finalTargetToken.safeTransferFrom(msg.sender, _transferAddress, sourceAmount);
        }
```

Reply:
fixed